### DISCUSSION TITLE
Report generation in Security Onion 2.3

### DISCUSSION THREAD

**Discussion by haleinfosec** [Security-Onion-Solutions/securityonion#1902] [Unsupported Versions] [answered]
Looking to generate reports based on alerts in Security Onion 2.3 - would like to know how to export alerts from Elasticsearch / Kibana in something like an excel spreadsheet so they can later be turned into PDF document. Open to any suggestions...

Is anyone currently creating reports from security onion data - how are you going about this? 

Thanks!

**Comment by weslambert** [comment, marked as answer]
We have report generation on the roadmap (https://github.com/Security-Onion-Solutions/securityonion/issues/936).  At the moment, you could try creating a custom dashboard in Grafana, which you could export, or you could enable Elastic Features with `so-features-enable`, restart services, and use the Kibana CSV Export URL to download a saved search as a CSV, or even create your own query and export it via a script.

Here is an example (URL initially copied from `Copy POST URL` and modified as necessary) that would download the last 24 hours of alerts, including `@timestamp`, `rule.name`, `destination.ip`, and `source.ip` : 

```
#!/bin/bash
KIBANA="http://localhost:5601"
VERSION=7.9.2
DL_PATH=$(curl -s -XPOST -H "kbn-version: $VERSION" "$KIBANA/api/reporting/generate/csv?jobParams=%28browserTimezone%3AAmerica%2FNew_York%2CconflictedTypesFields%3A%21%28%29%2Cfields%3A%21%28%27%40timestamp%27%2Crule.name%2Cdestination.ip%2Csource.ip%29%2CindexPatternId%3A%272289a0c0-6970-11ea-a0cd-ffa0f6a1bc29%27%2CmetaFields%3A%21%28_source%2C_id%2C_type%2C_index%2C_score%29%2CobjectType%3Asearch%2CsearchRequest%3A%28body%3A%28_source%3A%28includes%3A%21%28%27%40timestamp%27%2Crule.name%2Cdestination.ip%2Csource.ip%29%29%2Cdocvalue_fields%3A%21%28%28field%3A%27%40timestamp%27%2Cformat%3Adate_time%29%29%2Cquery%3A%28bool%3A%28filter%3A%21%28%28bool%3A%28minimum_should_match%3A1%2Cshould%3A%21%28%28match%3A%28event.dataset%3Aalert%29%29%29%29%29%2C%28range%3A%28%27%40timestamp%27%3A%28format%3Astrict_date_optional_time%2Cgte%3A%272020-11-09T20%3A45%3A37.772Z%27%2Clte%3A%272020-11-12T20%3A45%3A37.772Z%27%29%29%29%29%2Cmust%3A%21%28%29%2Cmust_not%3A%21%28%29%2Cshould%3A%21%28%29%29%29%2Cscript_fields%3A%28%29%2Csort%3A%21%28%28%27%40timestamp%27%3A%28order%3Adesc%2Cunmapped_type%3Aboolean%29%29%29%2Cstored_fields%3A%21%28%27%40timestamp%27%2Crule.name%2Cdestination.ip%2Csource.ip%29%2Cversion%3A%21t%29%2Cindex%3A%27%2A%3Aso-%2A%27%29%2Ctitle%3AMy_Alerts_Report%29" | jq -r .path | cut -d'/' -f7 )
echo "$DL_PATH"
JOB_STATUS=$(curl -s "$KIBANA/api/reporting/jobs/download/$DL_PATH")
while [[ "$JOB_STATUS" == "pending" || "$JOB_STATUS" == "processing" ]]; do
  echo "waiting..."
  JOB_STATUS=$(curl -s "$KIBANA/api/reporting/jobs/download/$DL_PATH")
done
echo "$JOB_STATUS" > results.csv
echo "Done!"
```

If you wanted to do this at a scheduled interval, expand on this, schedule via cron, and run non-interactively.  I'd be careful about performance impact, however, so it may be good to schedule at a non-peak time.

**Reply by stijnpieters** [reply]
Although the `so-features-enable` is exactly what someone would need, according to the license of elasticsearch you're not allowed to use the basic license (which `so-features-enable` activates) for production.

**Reply by dougburks** [reply]
@stijnpieters Where does the license say that?

**Reply by stijnpieters** [reply]
Correct me if I'm misinterpreting it, but does 2.2 in [the license](https://raw.githubusercontent.com/elastic/elasticsearch/master/licenses/ELASTIC-LICENSE.txt) not describe you can not use it in production?

**Reply by dougburks** [reply]
Are you referring to this?
```
2. COMMERCIAL SOFTWARE SOURCE CODE

  2.1 Limited License. Subject to the terms and conditions of Section 2.2 of
  this Agreement, Elastic hereby grants to You, AT NO CHARGE and for so long as
  you are not in breach of any provision of this Agreement, a limited,
  non-exclusive, non-transferable, fully paid up royalty free right and license
  to the Commercial Software in Source Code format, without the right to grant
  or authorize sublicenses, to prepare Derivative Works of the Commercial
  Software, provided You (i) do not hack the licensing mechanism, or otherwise
  circumvent the intended limitations on the use of Elastic Software to enable
  features other than Basic Features and Functions or those features You are
  entitled to as part of a Subscription, and (ii) use the resulting object code
  only for reasonable testing purposes.

  2.2 Restrictions. Nothing in Section 2.1 grants You the right to (i) use the
  Commercial Software Source Code other than in accordance with Section 2.1
  above, (ii) use a Derivative Work of the Commercial Software outside of a
  Non-production Environment, in any production capacity, on a temporary or
  permanent basis, or (iii) transfer, sell, rent, lease, distribute, sublicense,
  loan or otherwise make available the Commercial Software Source Code, in whole
  or in part, to any third party. Notwithstanding the foregoing, You may
  maintain a copy of the repository in which the Source Code of the Commercial
  Software resides and that copy may be publicly accessible, provided that you
  include this Agreement with Your copy of the repository.
```

My interpretation is that Section 2 has to do with access to the Commercial Software Source Code (the source code that's not part of the free and open source Elastic stack) and that Section 2.2(ii) talks about derivative works of that commercial software source code. So the license is trying to prevent folks from looking at the commercial software source code, creating a derivative work of that commercial software source code, and then running that derivative work in a production environment.

Security Onion has an official partnership with Elastic to redistribute not only the free and open source version of the Elastic Stack but also the Features version of the Elastic Stack. They are fully aware of Security Onion users running the Features version in production and they want them to continue doing that! :smile:

**Reply by stijnpieters** [reply]
That cleared it up for me, thanks a lot!

**Reply by achak17** [reply]
To automate Security Onion reports using Kibana, you can use the report automation tool Skedler. 

Refer to this blog to check out the easy steps to automate Security Onion reports: 
https://www.skedler.com/blog/automate-security-onion-reports-using-skedler/
